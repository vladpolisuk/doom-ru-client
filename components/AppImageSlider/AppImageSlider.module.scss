.slider_container {
	position: relative;
	display: flex;
	justify-content: center;
	align-items: center;
}

.images_container {
	position: relative;
	display: flex;
	flex-direction: row;
	overflow: hidden;
	width: 100%;
	height: 100%;
	border-radius: 5px;
	gap: 10px;
}

.button {
	display: flex;
	align-items: center;
	justify-content: center;
	width: 48px;
	height: 48px;
	border-radius: 50%;
	padding: 10px;
	position: absolute;
	z-index: 10;
	background-color: rgba(0, 0, 0, 0.3);
	color: #fff;
	font-size: 1.5rem;
	border: none;
	cursor: pointer;
	user-select: none;

	&:hover,
	&:focus-visible {
		background-color: #2929298f;
	}
}

.prev_button {
	left: 10px;
	padding-right: 12px;
}

.next_button {
	right: 10px;
	padding-left: 12px;
}

.images_carousel {
	display: flex;
	flex-direction: row;
	flex-wrap: wrap;
	gap: 15px;
	padding: 15px 0;

	&_item {
		padding: 0;
	}

	&_image {
		transition: all 200ms ease-in-out;
		border-radius: 5px;
		object-fit: cover;
		opacity: 0.6;
	}

	&_imageActive {
		transform: scale(1.05);
		opacity: 1;
	}
}
