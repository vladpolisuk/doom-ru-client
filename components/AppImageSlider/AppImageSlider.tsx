import clsx from 'clsx';
import Image from 'next/image';
import { CSSProperties, FC, memo, useState } from 'react';
import { FaChevronLeft, FaChevronRight } from 'react-icons/fa';
import { animated, useTransition } from 'react-spring';
import { BaseAppComponent } from '../../types/components';
import AppButton from '../AppButton/AppButton';
import s from './AppImageSlider.module.scss';

type Props = BaseAppComponent<HTMLDivElement> & {
	images: { url: string }[];
	imageStyles?: string;
	btnStyles?: string;
	width?: number;
	height?: number;
	withCarousel?: boolean;
};

/** ## App Image Slider
 * The common slider component in the application
 * @type `FC<AppImagesSlider>`
 * @memo `true`
 * @return `html:div`
 */
export const AppImageSlider: FC<Props> = memo(
	({ images, width = 800, height = 400, imageStyles = '', className, btnStyles = '', withCarousel, ...extra }) => {
		const [index, setIndex] = useState(0);
		const [direction, setDirection] = useState<'left' | 'right' | undefined>();

		const onPrevClick = () => {
			if (direction !== undefined) return;
			setDirection('left');
			setIndex(index => (index === 0 ? images.length - 1 : index - 1));
		};

		const onNextClick = () => {
			if (direction !== undefined) return;
			setDirection('right');
			setIndex(index => (index === images.length - 1 ? 0 : index + 1));
		};

		const transitions = useTransition(images[index], {
			from: { transform: `translate3d(${direction === 'left' ? '-' : ''}100%, 0, 0)` },
			enter: { transform: 'translate3d(0%, 0, 0)' },
			leave: { transform: `translate3d(${direction === 'right' ? '-' : ''}100%, 0, 0)` },
			onRest: () => {
				setDirection(undefined);
			}
		});

		const prevStyles = clsx(s.button, s.prev_button, btnStyles, 'active--scale', 'transition');
		const nextStyles = clsx(s.button, s.next_button, btnStyles, 'active--scale', 'transition');
		const carouselStyles = clsx(s.images_carousel, 'unlisted');
		const carouselItemStyles = clsx(s.images_carousel_item, 'transition');
		const styles: CSSProperties = {
			position: 'absolute',
			left: 0,
			top: 0,
			width: '100%',
			height: '100%'
		};

		return (
			<div>
				<div
					data-testid='app-image-slider'
					className={clsx(s.slider_container, className)}
					{...extra}>
					{images.length > 1 && (
						<AppButton
							resetStyles
							data-testid='app-image-slider-prev'
							className={prevStyles}
							onClick={onPrevClick}>
							<FaChevronLeft />
						</AppButton>
					)}

					<div className={clsx(s.images_container)}>
						{direction === undefined ? (
							<Image
								width={width}
								height={height}
								style={styles}
								alt='image'
								quality={100}
								className={imageStyles}
								src={images[index].url}
								data-testid='app-image-slider-image'
							/>
						) : (
							transitions((style, image) => (
								<animated.div
									key={image.url}
									style={{ ...style, ...styles }}>
									<Image
										width={width}
										height={height}
										alt='image'
										quality={100}
										className={imageStyles}
										src={image.url}
									/>
								</animated.div>
							))
						)}
					</div>

					{images.length > 1 && (
						<AppButton
							resetStyles
							data-testid='app-image-slider-next'
							className={nextStyles}
							onClick={onNextClick}>
							<FaChevronRight />
						</AppButton>
					)}
				</div>

				{withCarousel && images.length > 1 && (
					<ul
						data-testid='app-image-slider-carousel'
						className={carouselStyles}>
						{images.map((image, i: number) => (
							<li key={image.url}>
								<AppButton
									onClick={() => setIndex(i)}
									className={carouselItemStyles}
									color='transparent'>
									<Image
										alt='image'
										className={clsx(
											s.images_carousel_image,
											i === index && s.images_carousel_imageActive
										)}
										src={image.url}
										width={80}
										height={80}
									/>
								</AppButton>
							</li>
						))}
					</ul>
				)}
			</div>
		);
	}
);

AppImageSlider.displayName = 'AppImagesSlider';
