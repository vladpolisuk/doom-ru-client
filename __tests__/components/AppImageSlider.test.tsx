import '@testing-library/jest-dom';
import { fireEvent, render } from '@testing-library/react';
import { AppImageSlider } from '../../components/AppImageSlider/AppImageSlider';

const images = [
	{
		url: 'https://images.unsplash.com/photo-1493809842364-78817add7ffb?ixlib=rb-4.0.3&ixid=MnwxMjA3fDB8MHxwaG90by1wYWdlfHx8fGVufDB8fHx8&auto=format&fit=crop&w=1170&q=80'
	},
	{
		url: 'https://images.unsplash.com/photo-1522708323590-d24dbb6b0267?ixlib=rb-4.0.3&ixid=MnwxMjA3fDB8MHxwaG90by1wYWdlfHx8fGVufDB8fHx8&auto=format&fit=crop&w=1170&q=80'
	},
	{
		url: 'https://images.unsplash.com/photo-1538944570562-2c9cb7857097?ixlib=rb-4.0.3&ixid=MnwxMjA3fDB8MHxwaG90by1wYWdlfHx8fGVufDB8fHx8&auto=format&fit=crop&w=2070&q=80'
	},
	{
		url: 'https://images.unsplash.com/photo-1571843439991-dd2b8e051966?ixlib=rb-4.0.3&ixid=MnwxMjA3fDB8MHxwaG90by1wYWdlfHx8fGVufDB8fHx8&auto=format&fit=crop&w=1170&q=80'
	},
	{
		url: 'https://images.unsplash.com/photo-1541123437800-1bb1317badc2?ixlib=rb-4.0.3&ixid=MnwxMjA3fDB8MHxwaG90by1wYWdlfHx8fGVufDB8fHx8&auto=format&fit=crop&w=1170&q=80'
	}
];

describe('AppImageSlider', () => {
	it('should render', () => {
		const wrapper = render(<AppImageSlider images={images} />);
		const slider = wrapper.getByTestId('app-image-slider');
		expect(slider).toBeInTheDocument();
		expect(wrapper).toMatchSnapshot();
	});

	it('should render with carousel', () => {
		const wrapper = render(
			<AppImageSlider
				withCarousel
				images={images}
			/>
		);
		const carousel = wrapper.getByTestId('app-image-slider-carousel');
		expect(carousel.children.length).toBe(5);
		expect(carousel).toBeInTheDocument();
	});

	it('should change image on click', async () => {
		const wrapper = render(
			<AppImageSlider
				withCarousel
				images={images}
			/>
		);
		const next = wrapper.getByTestId('app-image-slider-next');
		const prev = wrapper.getByTestId('app-image-slider-prev');
		let image = wrapper.getByTestId('app-image-slider-image');
		expect(image.getAttribute('src')).toMatch(new RegExp(encodeURIComponent(images[0].url)));
		fireEvent.click(next);
		setTimeout(() => {
			image = wrapper.getByTestId('app-image-slider-image');
			expect(image.getAttribute('src')).toMatch(new RegExp(encodeURIComponent(images[1].url)));
			fireEvent.click(prev);
			setTimeout(() => {
				image = wrapper.getByTestId('app-image-slider-image');
				expect(image.getAttribute('src')).toMatch(new RegExp(encodeURIComponent(images[0].url)));
			}, 1000);
		}, 1000);
	});

	it('should change image on carousel click', async () => {
		const wrapper = render(
			<AppImageSlider
				withCarousel
				images={images}
			/>
		);
		const carousel = wrapper.getByTestId('app-image-slider-carousel');
		let image = wrapper.getByTestId('app-image-slider-image');
		expect(image.getAttribute('src')).toMatch(new RegExp(encodeURIComponent(images[0].url)));
		fireEvent.click(carousel.children[1]);
		setTimeout(() => {
			image = wrapper.getByTestId('app-image-slider-image');
			expect(image.getAttribute('src')).toMatch(new RegExp(encodeURIComponent(images[1].url)));
		}, 1000);
	});
});
