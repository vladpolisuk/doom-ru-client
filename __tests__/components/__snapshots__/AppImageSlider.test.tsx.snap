// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AppImageSlider should render 1`] = `
{
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div>
        <div
          class="slider_container"
          data-testid="app-image-slider"
        >
          <button
            class="button prev_button active--scale transition"
            data-testid="app-image-slider-prev"
          >
            <svg
              fill="currentColor"
              height="1em"
              stroke="currentColor"
              stroke-width="0"
              viewBox="0 0 320 512"
              width="1em"
              xmlns="http://www.w3.org/2000/svg"
            >
              <path
                d="M34.52 239.03L228.87 44.69c9.37-9.37 24.57-9.37 33.94 0l22.67 22.67c9.36 9.36 9.37 24.52.04 33.9L131.49 256l154.02 154.75c9.34 9.38 9.32 24.54-.04 33.9l-22.67 22.67c-9.37 9.37-24.57 9.37-33.94 0L34.52 272.97c-9.37-9.37-9.37-24.57 0-33.94z"
              />
            </svg>
          </button>
          <div
            class="images_container"
          >
            <img
              alt="image"
              class=""
              data-nimg="1"
              data-testid="app-image-slider-image"
              decoding="async"
              height="400"
              loading="lazy"
              src="/_next/image?url=https%3A%2F%2Fimages.unsplash.com%2Fphoto-1493809842364-78817add7ffb%3Fixlib%3Drb-4.0.3%26ixid%3DMnwxMjA3fDB8MHxwaG90by1wYWdlfHx8fGVufDB8fHx8%26auto%3Dformat%26fit%3Dcrop%26w%3D1170%26q%3D80&w=1920&q=100"
              srcset="/_next/image?url=https%3A%2F%2Fimages.unsplash.com%2Fphoto-1493809842364-78817add7ffb%3Fixlib%3Drb-4.0.3%26ixid%3DMnwxMjA3fDB8MHxwaG90by1wYWdlfHx8fGVufDB8fHx8%26auto%3Dformat%26fit%3Dcrop%26w%3D1170%26q%3D80&w=828&q=100 1x, /_next/image?url=https%3A%2F%2Fimages.unsplash.com%2Fphoto-1493809842364-78817add7ffb%3Fixlib%3Drb-4.0.3%26ixid%3DMnwxMjA3fDB8MHxwaG90by1wYWdlfHx8fGVufDB8fHx8%26auto%3Dformat%26fit%3Dcrop%26w%3D1170%26q%3D80&w=1920&q=100 2x"
              style="color: transparent; position: absolute; left: 0px; top: 0px; width: 100%; height: 100%;"
              width="800"
            />
          </div>
          <button
            class="button next_button active--scale transition"
            data-testid="app-image-slider-next"
          >
            <svg
              fill="currentColor"
              height="1em"
              stroke="currentColor"
              stroke-width="0"
              viewBox="0 0 320 512"
              width="1em"
              xmlns="http://www.w3.org/2000/svg"
            >
              <path
                d="M285.476 272.971L91.132 467.314c-9.373 9.373-24.569 9.373-33.941 0l-22.667-22.667c-9.357-9.357-9.375-24.522-.04-33.901L188.505 256 34.484 101.255c-9.335-9.379-9.317-24.544.04-33.901l22.667-22.667c9.373-9.373 24.569-9.373 33.941 0L285.475 239.03c9.373 9.372 9.373 24.568.001 33.941z"
              />
            </svg>
          </button>
        </div>
      </div>
    </div>
  </body>,
  "container": <div>
    <div>
      <div
        class="slider_container"
        data-testid="app-image-slider"
      >
        <button
          class="button prev_button active--scale transition"
          data-testid="app-image-slider-prev"
        >
          <svg
            fill="currentColor"
            height="1em"
            stroke="currentColor"
            stroke-width="0"
            viewBox="0 0 320 512"
            width="1em"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M34.52 239.03L228.87 44.69c9.37-9.37 24.57-9.37 33.94 0l22.67 22.67c9.36 9.36 9.37 24.52.04 33.9L131.49 256l154.02 154.75c9.34 9.38 9.32 24.54-.04 33.9l-22.67 22.67c-9.37 9.37-24.57 9.37-33.94 0L34.52 272.97c-9.37-9.37-9.37-24.57 0-33.94z"
            />
          </svg>
        </button>
        <div
          class="images_container"
        >
          <img
            alt="image"
            class=""
            data-nimg="1"
            data-testid="app-image-slider-image"
            decoding="async"
            height="400"
            loading="lazy"
            src="/_next/image?url=https%3A%2F%2Fimages.unsplash.com%2Fphoto-1493809842364-78817add7ffb%3Fixlib%3Drb-4.0.3%26ixid%3DMnwxMjA3fDB8MHxwaG90by1wYWdlfHx8fGVufDB8fHx8%26auto%3Dformat%26fit%3Dcrop%26w%3D1170%26q%3D80&w=1920&q=100"
            srcset="/_next/image?url=https%3A%2F%2Fimages.unsplash.com%2Fphoto-1493809842364-78817add7ffb%3Fixlib%3Drb-4.0.3%26ixid%3DMnwxMjA3fDB8MHxwaG90by1wYWdlfHx8fGVufDB8fHx8%26auto%3Dformat%26fit%3Dcrop%26w%3D1170%26q%3D80&w=828&q=100 1x, /_next/image?url=https%3A%2F%2Fimages.unsplash.com%2Fphoto-1493809842364-78817add7ffb%3Fixlib%3Drb-4.0.3%26ixid%3DMnwxMjA3fDB8MHxwaG90by1wYWdlfHx8fGVufDB8fHx8%26auto%3Dformat%26fit%3Dcrop%26w%3D1170%26q%3D80&w=1920&q=100 2x"
            style="color: transparent; position: absolute; left: 0px; top: 0px; width: 100%; height: 100%;"
            width="800"
          />
        </div>
        <button
          class="button next_button active--scale transition"
          data-testid="app-image-slider-next"
        >
          <svg
            fill="currentColor"
            height="1em"
            stroke="currentColor"
            stroke-width="0"
            viewBox="0 0 320 512"
            width="1em"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M285.476 272.971L91.132 467.314c-9.373 9.373-24.569 9.373-33.941 0l-22.667-22.667c-9.357-9.357-9.375-24.522-.04-33.901L188.505 256 34.484 101.255c-9.335-9.379-9.317-24.544.04-33.901l22.667-22.667c9.373-9.373 24.569-9.373 33.941 0L285.475 239.03c9.373 9.372 9.373 24.568.001 33.941z"
            />
          </svg>
        </button>
      </div>
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
